deps/thpool/thpool.c

```diff
@@ -391,108 +452,104 @@ thpool_stats redisearch_thpool_get_stats(redisearch_thpool_t *thpool_p) {
+  /* Use prctl instead to prevent using _GNU_SOURCE flag and implicit
+   * declaration */
   prctl(PR_SET_NAME, thread_name);
 #elif defined(__APPLE__) && defined(__MACH__)
   pthread_setname_np(thread_name);
 #else
-	LOG_IF_EXISTS("warning", "thread_do(): pthread_setname_np is not supported on this system")
+  LOG_IF_EXISTS(
+      "warning",
+      "thread_do(): pthread_setname_np is not supported on this system")
 #endif
 
-  /* Assure all threads have been created before starting serving */
-  redisearch_thpool_t* thpool_p = thread_p->thpool_p;
-  LOG_IF_EXISTS("verbose", "Creating background thread-%d", thread_p->id)
+  LOG_IF_EXISTS("verbose", "Creating background thread: %s", thread_name)
 
   /* Mark thread as alive (initialized) */
-  pthread_mutex_lock(&thpool_p->thcount_lock);
   thpool_p->num_threads_alive += 1;
-  pthread_mutex_unlock(&thpool_p->thcount_lock);
-
-  while (thpool_p->state & (THPOOL_KEEP_ALIVE | THPOOL_TERMINATE_WHEN_EMPTY)) {
-
-    bsem_wait(thpool_p->jobqueue.has_jobs);
-    LOG_IF_EXISTS("debug", "Thread-%d is running iteration", thread_p->id)
-    if (thpool_p->state & (THPOOL_KEEP_ALIVE | THPOOL_TERMINATE_WHEN_EMPTY)) {
-
-      pthread_mutex_lock(&thpool_p->thcount_lock);
-      size_t iter_id = thpool_p->num_threads_working++;
-      pthread_mutex_unlock(&thpool_p->thcount_lock);
-
-      /* Read job from queue and execute it */
-      void (*func_buff)(void*);
-      void* arg_buff;
-      job* job_p = priority_queue_pull(&thpool_p->jobqueue, iter_id);
-      if (job_p) {
-        func_buff = job_p->function;
-        arg_buff = job_p->arg;
-        func_buff(arg_buff);
-        rm_free(job_p);
-      }
+  ThreadCtx thread_ctx = {.thread_state = THREAD_RUNNING};
 
-      pthread_mutex_lock(&thpool_p->thcount_lock);
-      thpool_p->num_threads_working--;
-      if (job_p) thpool_p->total_jobs_done++;
-      if (thpool_p->num_threads_working == 0) {
-	      LOG_IF_EXISTS("debug", "All threads are idle")
-	      pthread_cond_signal(&thpool_p->threads_all_idle);
-      }
-      if (thpool_p->state == THPOOL_TERMINATE_WHEN_EMPTY && priority_queue_len(&thpool_p->jobqueue) == 0) {
-        LOG_IF_EXISTS("verbose", "Job queue is empty - terminating thread %d", thread_p->id);
-        thpool_p->state = THPOOL_UNINITIALIZED;
+  while (true) {
+    LOG_IF_EXISTS("debug", "Thread %s is running iteration", thread_name)
+
+    /* Read job from queue and execute it */
+    PriorityJobCtx job_ctx = priority_queue_pull(&thpool_p->jobqueue);
+    job *job_p = job_ctx.job;
+    void *arg = job_p->arg;
+
+    AdminJobArg admin_job_arg = {0};
+    if (job_ctx.is_admin) {
+      admin_job_arg.arg = arg;
+      admin_job_arg.thread_ctx = &thread_ctx;
+      arg = &admin_job_arg;
+    }
+
+    job_p->function(arg);
+    rm_free(job_p);
+
+    // Both variables are atomic, so we can do this without a lock.
+    thpool_p->total_jobs_done += !job_ctx.is_admin;
```

Comment:
```suggestion
    if (!job_ctx.is_admin) thpool_p->total_jobs_done++;
```

Reply:
in the C Standard (ISO/IEC 9899:2018):

Section 6.5.3.3 Unary arithmetic operators: "The result of the logical negation operator ! is 0 if the value of its operand compares unequal to 0, 1 if the value of its operand compares equal to 0. The result has type int."